// raml-outline/dist/node_details/detailsInterfaces.js
"use strict";
/**
 * Type of details item
 */
(function (DetailsItemType) {
    DetailsItemType[DetailsItemType["ROOT"] = 0] = "ROOT";
    DetailsItemType[DetailsItemType["CATEGORY"] = 1] = "CATEGORY";
    DetailsItemType[DetailsItemType["CHECKBOX"] = 2] = "CHECKBOX";
    DetailsItemType[DetailsItemType["JSONSCHEMA"] = 3] = "JSONSCHEMA";
    DetailsItemType[DetailsItemType["XMLSCHEMA"] = 4] = "XMLSCHEMA";
    DetailsItemType[DetailsItemType["MARKDOWN"] = 5] = "MARKDOWN";
    DetailsItemType[DetailsItemType["SELECTBOX"] = 6] = "SELECTBOX";
    DetailsItemType[DetailsItemType["MULTIEDITOR"] = 7] = "MULTIEDITOR";
    DetailsItemType[DetailsItemType["TREE"] = 8] = "TREE";
    DetailsItemType[DetailsItemType["STRUCTURED"] = 9] = "STRUCTURED";
    DetailsItemType[DetailsItemType["TYPEDISPLAY"] = 10] = "TYPEDISPLAY";
    DetailsItemType[DetailsItemType["TYPESELECT"] = 11] = "TYPESELECT";
    DetailsItemType[DetailsItemType["JSONEXAMPLE"] = 12] = "JSONEXAMPLE";
    DetailsItemType[DetailsItemType["XMLEXAMPLE"] = 13] = "XMLEXAMPLE";
    DetailsItemType[DetailsItemType["ATTRIBUTETEXT"] = 14] = "ATTRIBUTETEXT";
})(exports.DetailsItemType || (exports.DetailsItemType = {}));
var DetailsItemType = exports.DetailsItemType;
